crm_compassion/models/weekly_demand.py

```diff
@@ -99,61 +113,104 @@ def _compute_demand_events(self):
         ])
         allocate_per_week = int(self.env['ir.config_parameter'].get_param(
             'crm_compassion.number_children_web'))
-        return allocate_per_week - (float(web_sponsored) / 52.0)
+        return allocate_per_week - (float(web_sponsored) / STATS_DURATION)
 
     @api.model
     def _default_unsponsored_ambassador(self):
         """ Compute average of unsponsored children from ambassadors since
         one year.
         """
-        start_date = datetime.today() - timedelta(weeks=52)
+        start_date = datetime.today() - timedelta(weeks=STATS_DURATION)
         ambass_sponsored = self.env['recurring.contract'].search_count([
             ('origin_id.type', '=', 'partner'),
             ('origin_id.partner_id', '!=', False),
             ('start_date', '>=', fields.Date.to_string(start_date))
         ])
         allocate_per_week = int(self.env['ir.config_parameter'].get_param(
             'crm_compassion.number_children_ambassador'))
-        return allocate_per_week - (float(ambass_sponsored) / 52.0)
+        return allocate_per_week - (float(ambass_sponsored) / STATS_DURATION)
+
+    @api.depends('week_start_date')
+    @api.multi
+    def _compute_resupply_sub(self):
+        """ Compute SUB resupply. """
+        sub_average = self._default_demand_sub()
+        today = datetime.today()
+        rejected_sub = self.env['recurring.contract'].search([
+            ('origin_id.type', '=', 'sub'),
+            ('start_date', '>=',
+             datetime.today() - timedelta(weeks=STATS_DURATION)),
+            ('end_date', '!=', None)
+        ]).filtered(
+            lambda s: ((
+                           fields.Date.from_string(s.end_date) -
+                           fields.Date.from_string(s.start_date)
+                       ).days <= SUB_DURATION)
+        )
+        sub_reject_average = len(rejected_sub) / STATS_DURATION
+        for week in self:
+            start_date = fields.Datetime.from_string(
+                week.week_start_date) - timedelta(days=SUB_DURATION)
+            if start_date <= today:
+                sub = self.env['recurring.contract'].search_count([
+                    ('origin_id.type', '=', 'sub'),
+                    ('start_date', '>=', start_date),
+                    ('start_date', '<=', fields.Date.from_string(
+                        week.week_end_date) - timedelta(days=SUB_DURATION))
+                ])
+                week.resupply_sub = sub * (sub_reject_average / sub_average)
+            else:
+                week.resupply_sub = sub_reject_average
 
     @api.model
     def _default_cancellation(self):
         """ Compute average of sponsor cancellations since one year. """
-        start_date = datetime.today() - timedelta(weeks=52)
+        start_date = datetime.today() - timedelta(weeks=STATS_DURATION)
         cancellations = self.env['recurring.contract'].search_count([
             ('state', '=', 'terminated'),
```

Comment:
Can you add a condition `('origin_id.type', '!=', 'sub')` ? To avoid counting the sub resupply in cancellations.